### DISCUSSION TITLE
Database Implementations

### DISCUSSION THREAD

**Discussion by m1sterc001guy** [fedimint/fedimint#1836] [General]
This is a thread for discussing the pros/cons of different database implementations that can be used in Fedimint.

## Implemented
**RocksDb**
This is the default database implementation that is supported today for the server and client. It is an embeded key/value store written in C++. RocksDb is a fork of LevelDB and underneath uses LSM trees for storing data. Supports snapshot isolation and optimistic transactions. RocksDb optimistic transactions supports multiple writes, as long as the keys they are modifying do not conflict. Conflicts are checked at commit time. Does not have great support for asynchronous reads/writes.

**Memory Database**
Custom in-memory database built for Fedimint testing using BTreeMaps. Should not be used in production as it doesn't persist data and doesn't support full snapshot isolation.

**SQLite**
Embedded relational database written in C. In Fedimint, our implementation uses a single table and presents a key/value interface to developers. SQLite is a reliable database and is used across many different projects. It supports snapshot isolation, but does not support optimistic concurrency control. This is somewhat problematic today because there are certain parts of the Fedimint codebase that assume optimistic concurrency control for error handling. SQLite only supports one writer and many readers, but the impact of a single writer can be mitigated by turning on the WAL. There is a separate branch from 2019 to add optimistic transactions to SQLite but it has not been merged upstream.

We are using the sqlx library for sqlite support, which unfortunately has lots of dependencies. However, sqlx implements an async/await interface for interacting with sqlite.

## Potential Options
**SledDB**
Sled is an embedded key/value store written in pure rust. It supports serializable transactions and optimistic concurrency control. Has support for async reads and writes. Earlier versions of Fedimint had support for Sled, but it was removed because transactions in sled occur within a closure, which makes it difficult to work with in Fedimint's code (we essentially needed to implement an in-memory caching layer so that transactions using sled could read their own writes). Sled is younger and newer than the other implementations. 

**LMDB**
Memory mapped embedded key value store written in C. Based on BerkleyDB. Has rust bindings. Supports snapshot isolation with multiple readers, but only one writer at time (similar to sqlite) and uses locking to achieve this. Unlike other options, this implementation can support multiple processes reading the db. Does not require transaction log.

**ReDB**
Key/value store written in Rust. Supports serializable transactions, multiple readers, but only one writer. Uses copy-on-write to implement MVCC. 

**AgateDB**
Very new key/value store written in Rust. Project is porting BadgerDB to rust. Not a ton of documentation, looks like it is mainly being developed for a specific project. Doesn't support async IO yet.

**Comment by dpc** [comment]
Memory Database is OK for simple unit-tests, but even then it's probably better to just use rocksdb. The different in performance is neglible, I'm guessing, and testing a real thing is always better than some mock.

I don't really see a a practical way for us to support multiple databases. It just makes all the code-paths less tested, especially for non-default one, which renders it not something people should use in a mission critical software like Fedimint.

So we should probably accept the fact that it will be rocksdb for everything for foreseeable future. My only problem with rocksdb is that it's C++ and so has a really meh cross-compilation story, and usually I have to deal with it. :D

**Comment by justinmoon** [comment]
Some discussion on [SurrealDB](https://surrealdb.com/) in [this closed issue](https://github.com/fedimint/fedimint/issues/3398#issuecomment-1767239070)
